### PR TITLE
Add 2fa to registration flow

### PR DESCRIPTION
**Pull request by frankieroberto**
This requires users to complete a mobile phone number verification step (entering a code received via text message) when completing their account details after being invited to use the service.

Because entering name, password and mobile number is a previous step, the user can potentially be left in a state where they have set those details but haven't verified their mobile number. In this (hopefully rare) state, they:

* Can still go back and start their account setup process again (from the link in the e-mail), provided it’s within the expiry period.
* Cannot login - if they do, they get a generic error message (this prevents a user from ending up in a state where they have an active account but can't sign in because their mobile number is invalid).

See https://trello.com/c/Huo9xSPj/389-2fa-on-registration-page

## Implementation approach

To keep track of whether the mobile number has been verified (via 2FA), a new `mobile_number_verified` boolean attribute is added to the User model. This is set to true after a user completes 2FA. The sign in page checks that this is true. Until it becomes `true`, the invitation to set up an account remains valid.

## Alternative approaches 

* #398 doesn’t save the name, password and mobile number to the database at all until the 2FA is complete. Instead it saves it to the redis session store.
* #404 as above, but saving to the Rails cache instead (which I think is also backed by Redis).

## Checklist:
- [x] Have you documented your changes in the pull request description?
- [x] Does the change present any security considerations?
- [ ] Has acceptance criteria been tested by a peer?
- [ ] Has the CHANGELOG been updated? (If change is worth telling users about.)

### CODE REVIEW

File: psd-web/db/migrate/20200320144542_add_mobile_number_verified_to_users.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class AddMobileNumberVerifiedToUsers < ActiveRecord::Migration[5.2]
+  def change
+    safety_assured do
+      add_column :users, :mobile_number_verified, :boolean, default: false, null: false
```
Comment (slawosz) [inline psd-web/db/migrate/20200320144542_add_mobile_number_verified_to_users.rb]: How about making this `datetime`, this way we would know when it was verified?

File: psd-web/app/controllers/users/sessions_controller.rb [hunk 2]
```diff
@@ -18,6 +18,14 @@ def create
 
       self.resource = warden.authenticate(auth_options)
 
+
```
Comment (scruti) [inline psd-web/app/controllers/users/sessions_controller.rb]: :scissors:

File: psd-web/spec/features/creating_an_account_from_an_invitation_spec.rb [hunk 3]
```diff
@@ -14,6 +14,28 @@
...
+    expect_to_be_on_the_homepage
+
+    click_link "Sign in to your account"
+
+    fill_in "Email address", with: invited_user.email
+    fill_in "Password", with: "testpassword123@"
+    click_on "Continue"
+
+    # Skips 2FA as cookie was set to not require
+    # 2FA for 7 days.
```
Comment (scruti) [inline psd-web/spec/features/creating_an_account_from_an_invitation_spec.rb]: Nice explanation :clap:

File: psd-web/spec/requests/two_factor_authentication_spec.rb [hunk 4]
```diff
@@ -67,6 +69,11 @@
           follow_redirect!
           expect(response.body).to include("Sign out")
         end
+
+        it "marks the mobile number as verified" do
+          submit_2fa
+          expect(user.reload.mobile_number_verified).to be true
```
Comment (scruti) [inline psd-web/spec/requests/two_factor_authentication_spec.rb]: Suggestion: 
Suggestion: As @StupidCodeFactory would say... Rspec provides nice helpers for this 
```suggestion
          expect(user.reload).to be_mobile_number_verified
```

File: psd-web/spec/requests/user_completes_registration_spec.rb [hunk 5]
```diff
@@ -101,19 +113,23 @@
         expect(user).not_to be_account_activated
       end
 
-      it "redirects to the root path" do
-        expect(response).to redirect_to(root_path)
+      it "does not set the mobile number as verified" do
+        expect(user.mobile_number_verified).to be false
```
Comment (scruti) [inline psd-web/spec/requests/user_completes_registration_spec.rb]: Suggestion (showing the possibility to be considered more than requesting any change):
```suggestion
        expect(user).not_to be_mobile_number_verified
```

File: psd-web/spec/requests/two_factor_authentication_spec.rb (hunk 4 above)
Comment (frankieroberto) [reply psd-web/spec/requests/two_factor_authentication_spec.rb]: I personally like the rspec helpers, but only where they make linguistic sense. "be mobile number verified" feels a bit awkward and ungrammatical to me?

File: psd-web/db/migrate/20200320144542_add_mobile_number_verified_to_users.rb (hunk 1 above)
Comment (frankieroberto) [reply psd-web/db/migrate/20200320144542_add_mobile_number_verified_to_users.rb]: Could do. I can’t immediately think of a reason why we'd need to know the time it was verified, but perhaps it’d be useful.

**Pull request merged**